agent/grpc-external/limiter/limiter.go

```diff
@@ -0,0 +1,254 @@
+			return
+		}
+	}
+}
+
+// SetMaxSessions controls the maximum number of concurrent sessions. If it is
+// lower, randomly-selected sessions will be drained.
+func (l *SessionLimiter) SetMaxSessions(max uint32) {
+	atomic.StoreUint32(&l.max, max)
+
+	// Send on wakeCh without blocking if the Run loop is busy. wakeCh has a
+	// buffer of 1, so no triggers will be missed.
+	select {
+	case l.wakeCh <- struct{}{}:
+	default:
+	}
+}
+
+// SetDrainRateLimit controls the rate at which excess sessions will be drained.
+func (l *SessionLimiter) SetDrainRateLimit(limit rate.Limit) {
+	l.drainLimiter.SetLimit(limit)
+}
+
+// BeginSession begins a new session, or returns ErrCapacityReached if the
+// concurrent session limit has been reached.
+//
+// It is the session-holder's responsibility to:
+//
+//	1. Call End on the session when finished.
+//	2. Receive on the session's Terminated channel and exit (e.g. close the gRPC
+//	   stream) when it is closed.
+func (l *SessionLimiter) BeginSession() (Session, error) {
+	if !l.hasCapacity() {
+		return nil, ErrCapacityReached
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.createSessionLocked(), nil
+}
+
+// Note: hasCapacity is *best effort*. As we do not hold l.mu it's possible that:
+//
+//	- max has changed by the time we compare it to inFlight.
+//	- inFlight < max now, but increases before we create a new session.
+//
+// This is acceptable for our uses, especially because excess sessions will
+// eventually be drained.
+func (l *SessionLimiter) hasCapacity() bool {
+	max := atomic.LoadUint32(&l.max)
+	if max == Unlimited {
+		return true
+	}
+
+	cur := atomic.LoadUint32(&l.inFlight)
+	return max > cur
+}
+
+// Note: overCapacity is *best effort*. As we do not hold l.mu it's possible that:
+//
+//	- max has changed by the time we compare it to inFlight.
+//	- inFlight > max now, but decreases before we terminate a session.
+func (l *SessionLimiter) overCapacity() bool {
+	max := atomic.LoadUint32(&l.max)
+	if max == Unlimited {
+		return false
+	}
+
+	cur := atomic.LoadUint32(&l.inFlight)
+	return cur > max
+}
+
+func (l *SessionLimiter) terminateSession() {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	idx := rand.Intn(len(l.sessionIDs))
+	id := l.sessionIDs[idx]
+	l.sessions[id].terminate()
+	l.deleteSessionLocked(idx)
+}
+
+func (l *SessionLimiter) createSessionLocked() *session {
+	session := &session{
+		l:      l,
+		id:     l.maxSessionID,
+		termCh: make(chan struct{}),
+	}
+
+	l.maxSessionID++
+	l.sessionIDs = append(l.sessionIDs, session.id)
+	l.sessions[session.id] = session
+
+	atomic.AddUint32(&l.inFlight, 1)
+
+	return session
+}
+
+func (l *SessionLimiter) deleteSessionLocked(idx int) {
+	delete(l.sessions, l.sessionIDs[idx])
+
+	// Note: it's important that we preserve the order here (which most allocation
+	// free deletion tricks don't) because we binary search the slice.
+	l.sessionIDs = append(l.sessionIDs[:idx], l.sessionIDs[idx+1:]...)
+
+	atomic.AddUint32(&l.inFlight, ^uint32(0))
```

Comment:
🎨 (code style / non-blocking)

Is there any reason we couldn't use a signed integer for these counts? Then this could be `atomic.AddInt32(&l.inFlight, -1)` which might be more immediately obvious to future readers what is happening.

Reply:
I went with unsigned because they shouldn't ever be negative (so it felt best make it impossible) but I'm happy to be pragmatic about this for readability!